### PR TITLE
refactor: raise test coverage to 90%+

### LINKED ISSUE #23: Close Assigned Issues
We do not want assigned issues to render on work.ubq.fi 

The simple solution would be to close them (that way they would not be categorized as an "open" task)

The more complex solution would be to have special logic in the UI for the `unavailable` label but I feel that closing as not planned might be better? Then we can close as complete if it is actually closed as complete? 

> > However, the original issue was resolved according to the spec here and that was just to remove the `unavailable` bounties from the devpool repo was it not? 
> 
> You're right. I read the following as "closing" the issue which means that it would not render in the work.ubq.fi UI
> 
> > if bounty is unavailable (i.e. assigned to some bounty hunter) and exists in the list with the state open then close such bounty
> 
> With an emphasis on 
> 
> > then close such bounty
> 
> So we need to file a new issue and get a new pull request to actually close them?

_Originally posted by @pavlovcik in https://github.com/ubiquity/devpool-directory-bounties/issues/11#issuecomment-2017567237_

### PR DESCRIPTION
**Pull request by Keyrxng**
Resolves https://github.com/ubiquity/devpool-directory-bounties/issues/23

The reason I've done this was because I was tired of things passing when being QA'd them then breaking a few hours later.

- raised test coverage to 90%
- tested all intended scenarios
- added the `Unavailable` label only to issues that are assigned-closed but not yet merged, removing when they are reopened-unassigned
- fixed the repoUrl parsing `Opt.in`/ `Opt.out`
- fixed the statistics to work with the new setup

(would have pushed it sooner but my actions stopped worked overnight and had to open a ticket)

- https://github.com/ubq-testing/devpool-directory/issues

image

### CODE REVIEW

File: helpers/github.ts
Comment (0x4007) [inline helpers/github.ts]: "opts" generally implies options like parameters for a function which can be misleading. I would stick with a derivative of "opt"

File: helpers/twitter.ts
Comment (0x4007) [inline helpers/twitter.ts:38]: I had no idea that this was implemented. I wonder if it actually works as expected.

**Review by 0x4007 — COMMENTED** [review]
It's a very large pull which makes it more dangerous to review from mobile and approve. 

Seems decent overall. 

In order to reduce the burden on reviewers, consider cherry picking changes and being more precise with pulls to linking issues. 

For example in the past I've done similar volume of changes, but I generally would branch off and open separate pulls based on eachother. 

Also at risk of being a bit bureaucratic, if you filed issues for the problems you are seeking to solve, we could have set pricing! But proactive fixes with a focus on fast delivery is always appreciated. Thanks!

**Comment by Keyrxng** [comment]
if review hasn't yet begun I'd willing be to attempt to split it into distinct PRs attached to respective issues if it makes things easier for everyone involved

- Issue: Raise test coverage > PR: add new test file and adjust old test files (using mocks with untouched `src/*`)
- Issue: Refactor `src/*`, ...list of fixes > PR: merge mocks into src
- Issue: Refactor statistics > PR: (I'd redo-statistics altogether so that it doesn't rely on `Unavailable` and that can stop being applied inside `getDevpoolIssues()` and removed altogether)

I think makes sense but if someone has a better suggestion I'll do that

<hr/>

The test file I've added is monstrous, I was going test by test just using `test.only/describe.only` but it should probably be split into one file per describe block or something right?

**Comment by Keyrxng** [comment]
Also check this run out with 25min runtime lmao

- https://github.com/ubq-testing/devpool-directory/actions/runs/8460262630

They would just sit in `Status: Queued` indefinitely until I manually cancelled the run, the odd part being that it would finish execution in normal time but it would never register as having left a `Queued` state. I could verify this because I was able to access the `twitterMap.json` output file a min or after it entering a `Queued` state which only happens as the last step.

Everything was in limbo CI, Cron, push, pr. I thought it was because of changes I had made but unfortunately not so I've no idea what caused it. I opened a ticket after a couple hours went to sleep and when I woke it had been resolved/resolved itself

File: helpers/github.ts
Comment (gentlementlegen) [inline helpers/github.ts]: Would be nice to change the `{ [key: string]: string }` to something like `Record<string, string>`

File: helpers/github.ts
Comment (gentlementlegen) [inline helpers/github.ts]:
```suggestion
```

File: index.ts
Comment (gentlementlegen) [inline index.ts]:
```suggestion
  // Calculate total rewards from open issues
```

File: tests/twitter.test.ts
Comment (gentlementlegen) [inline tests/twitter.test.ts]:
```suggestion
```

**Review by gentlementlegen — COMMENTED** [review]
Tested locally, all worked fine, thanks for the thorough testing. A few fixes and I am good with this.

**Comment by 0x4007** [comment]
@gentlementlegen when I make small "code style" adjustments like you did, I directly commit the suggestion (there is a button on the GitHub UI) I think that you should just commit them and we can immediately merge etc. Looks like I can't commit your suggestions for some reason though.

**Comment by gentlementlegen** [comment]
@pavlovcik yep I tried to do it that way? I preferred that @Keyrxng double checked and was committing himself the small changes to be sure to not break anything. Will try to push them

File: tests/twitter.test.ts
Comment (Keyrxng) [reply tests/twitter.test.ts]: I did this for two reasons the first was occasionally it was failing and it was because it was being imported but the not doing anything with it was likely the cause of it failing and the second was to silence the unused var error.

I tried it again without doing anything with the import and it passed each time but I've replaced it with `() => twitterHelper`. 1. to stop the unused error and 2, just in case doing nothing with the import causes nothing to happen in the test

**Comment by Keyrxng** [comment]
> Tested locally, all worked fine, thanks for the thorough testing. A few fixes and I am good with this.

glad to hear it

**Comment by Keyrxng** [comment]
Would be great to see this merged if it's ready to go, man's broke 😂

File: mocks/handlers.ts
Comment (gentlementlegen) [inline mocks/handlers.ts]: Would preferably not have this disabled here, what is the reason?

File: mocks/mock-issue-handlers.ts
Comment (gentlementlegen) [inline mocks/mock-issue-handlers.ts]: These two logs seem redundant

File: helpers/github.ts
Comment (gentlementlegen) [inline helpers/github.ts]: Why not `Object.values` if the key is not used?

File: helpers/github.ts
Comment (gentlementlegen) [inline helpers/github.ts]: Should use `===` to avoid any type coercion

File: helpers/twitter.ts
Comment (gentlementlegen) [reply helpers/twitter.ts:38]: After looking at the code, I have a hard time to believe it does. It seems to keep track of the posted tweets through a `json` file which I don't believe is reliable or even persisted.

File: helpers/github.ts
Comment (gentlementlegen) [inline helpers/github.ts:427]: This function truly is massive and would benefit to be broken down into smaller parts

File: helpers/github.ts
Comment (gentlementlegen) [inline helpers/github.ts:417]: How can this file be persisted?

File: mocks/handlers.ts
Comment (Keyrxng) [reply mocks/handlers.ts]: resolved by typing the label

File: helpers/github.ts
Comment (Keyrxng) [reply helpers/github.ts]: for of guarantees sequential execution as far as I understand which is preferred here

File: helpers/github.ts
Comment (Keyrxng) [reply helpers/github.ts]: I read somewhere that `===` isn't necessary with TS but I went and did a bit more reading just to be sure before commenting and I was wrong to have used `==`, my mistake

File: helpers/github.ts
Comment (Keyrxng) [reply helpers/github.ts:417]: It could be written to GH the same way that the total-rewards are I guess? I don't think twitter is actually in use tho

**Comment by Keyrxng** [comment]
there is no issue directly tied to raising the test coverage, should there be? I think so given some of the advice pav has given me but I'm not sure in this case.

Plus `Resolves https://github.com/ubiquity/devpool-directory-bounties/issues/23` is currently assigned to another user who has closed their pr a few weeks ago and I believe that issue is resolved, so I think it would make sense to have an issue for test coverage but again, I'm unsure.

File: helpers/github.ts
Comment (gentlementlegen) [reply helpers/github.ts:417]: I don't think so either But since this file is not committed it just gets rewritten every time so I really don't get how it works.

File: helpers/github.ts
Comment (gentlementlegen) [reply helpers/github.ts]: Not sure to understand why `entries` would keep it and not `values`.
```JavaScript
const obj = {
  foo: "bar",
  bar: "baz",
  baz: "boz",
};

for (const [, value] of Object.entries(obj)) {
  console.log(value);
}

console.log("-------");

for (const value of Object.values(obj)) {
  console.log(value);
}
```
would output
```text
bar
baz
boz
-------
bar
baz
boz
```
Objects are somehow predictable since ES2015 (see https://stackoverflow.com/questions/5525795/does-javascript-guarantee-object-property-order/38218582#38218582) but not guaranteed. So if that is a concern it should be changed to use an array or a map.

Comment (Keyrxng) [reply helpers/github.ts]: my mistake, I thought you meant to use

```js
Object.values().map // or .forEach()
```

File: helpers/github.ts
Comment (Keyrxng) [reply helpers/github.ts:417]: it's written into the action run outputs but as far as what happens with it after the fact or what is intended for it I'm not sure either.

Comment (gentlementlegen) [reply helpers/github.ts:417]: I thought you authored that code that's why I asked. I believe this would be for another task.

File: helpers/twitter.ts
Comment (Keyrxng) [reply helpers/twitter.ts:38]: is this holding back the PR?

**Comment by Keyrxng** [comment]
I've updated the referenced issues. But again, I'm unsure what to do about either creating a new issue for these tests specifically or removing the assignee from the linked issue, assigning me and closing that with this PR? Cheers

**Review by gentlementlegen — APPROVED** [review]

**Comment by Keyrxng** [comment]
Great, just waiting for final approval from yourself @0x4007. Love to see it, and ty for the review

**Pull request merged**
